From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mechoriet <kevinworm92@gmail.com>
Date: Mon, 19 Dec 2022 23:48:44 +0100
Subject: [PATCH] Implement SpawnerPreSpawnEvent with check to see if there are
 listeners for it so we can limit the amount of events fired


diff --git a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
index e54f7edbd1d834391f6760ea7b82cf6af4ddf994..c84d4172e217db987271f2e26f3a8befefac288c 100644
--- a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
+++ b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
@@ -5,6 +5,7 @@ import java.util.Iterator;
 import java.util.List;
 
 // CraftBukkit start
+import net.techcable.tacospigot.event.entity.SpawnerPreSpawnEvent;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.bukkit.event.entity.SpawnerSpawnEvent;
@@ -91,6 +92,17 @@ public abstract class MobSpawnerAbstract {
                 boolean flag = false;
 
                 for (int i = 0; i < this.spawnCount; ++i) {
+                    if(SpawnerPreSpawnEvent.getHandlerList().getRegisteredListeners().length > 0) {
+                        // TacoSpigot start
+                        SpawnerPreSpawnEvent event = new SpawnerPreSpawnEvent(new org.bukkit.Location(this.a().getWorld(), blockposition.getX(), blockposition.getY(), blockposition.getZ()), org.bukkit.entity.EntityType.fromName(this.getMobName()));
+                        this.a().getServer().getPluginManager().callEvent(event);
+
+                        if (event.isCancelled()) {
+                            flag = true;
+                            break;
+                        }
+                        // TacoSpigot end
+                    }
                     Entity entity = EntityTypes.createEntityByName(this.getMobName(), this.a());
 
                     if (entity == null) {
